### PR TITLE
GODRIVER-559 Add ServerSelectionError type and improve ConnectionError format/messages 

### PR DESCRIPTION
**Pull request by KritiRav**
https://jira.mongodb.org/browse/GODRIVER-559

If there's a configuration error that prevents any connection handshakes from succeeding (e.g. configuring a Client with no TLS enabled when the server requires TLS connections), operations will fail with a server selection timeout error like: 

`server selection error: server selection timeout, current topology: { Type: Unknown, Servers: [{ Addr: localhost:27017, Type: Unknown, State: Connected, Average RTT: 0, Last error: connection() : connection(localhost:27017[-10]) incomplete read of message header: EOF },] `

The error does not indicate that we have failed during a handshake connection and the last error provided (incomplete read of message header) is not helpful. The fix will look like: 

`server selection error: server selection timeout, current topology: { Type: Unknown, Servers: [{ Addr: localhost:27017, Type: Unknown, Average RTT: 0, Last error: connection() error occured during connection handshake: connection(localhost:27017[-5]) Socket was unexpectedly closed: EOF }, ] }`

There is also a ServerSelectionError type that was added which provides more information when context expires during server selection. This error will return:

`server selection error: context deadline exceeded, current topology: { Type: Unknown, Servers: [{ Addr: localhost:27017, Type: Unknown, Average RTT: 0, Last error: connection(localhost:27017[-5]) Socket was unexpectedly closed: EOF }, ] }`

Previously, the driver would return `ctx.Err() (i.e. context deadline exceeded)`

### CODE REVIEW

File: x/mongo/driver/topology/topology.go [hunk 1]
```diff
@@ -444,9 +440,9 @@ func (t *Topology) selectServerFromSubscription(ctx context.Context, subscriptio
 	for {
 		select {
 		case <-ctx.Done():
-			return nil, ctx.Err()
+			return nil, ServerSelectionError{Wrapped: ctx.Err(), Desc: t.Description()}
```
Comment (iwysiu) [inline x/mongo/driver/topology/topology.go:443]: Why is this one a ServerSelectionError when the original one wasn't?

File: x/mongo/driver/description/topology.go [hunk 2]
```diff
@@ -91,3 +93,12 @@ func (t Topology) DiffHostlist(hostlist []string) HostlistDiff {
 
 	return diff
 }
+
+// String implements the Stringer interface
+func (t Topology) String() string {
+	serversStr := ""
```
Comment (divjotarora) [inline x/mongo/driver/description/topology.go:99]: nit: If you want the zero value of a string, you can initialize this as `var serversStr string`

File: x/mongo/driver/topology/topology.go (hunk 1 above)
Comment (iwysiu) [reply x/mongo/driver/topology/topology.go:443]: oops, somehow missed your original comment

File: x/mongo/driver/topology/connection.go [hunk 3]
```diff
@@ -292,10 +294,16 @@ func (c *connection) readWireMessage(ctx context.Context, dst []byte) ([]byte, e
 	if err != nil {
 		// We closeConnection the connection because we don't know if there are other bytes left to read.
 		c.close()
+		var message string
```
Comment (divjotarora) [inline x/mongo/driver/topology/connection.go:297]: nit: We generally try to avoid if/else statements. You can initialize `message` to a default and then overwrite it in a condition. The general case is "incomplete read of message header" so this would look like:

```
message := "incomplete read of message header"
if err == io.EOF {
    message = "Socket was unexpectedly closed"
}
```

File: x/mongo/driver/topology/connection.go [hunk 4]
```diff
@@ -292,10 +294,16 @@ func (c *connection) readWireMessage(ctx context.Context, dst []byte) ([]byte, e
...
+		if err == io.EOF {
+			message = "Socket was unexpectedly closed"
```
Comment (divjotarora) [inline x/mongo/driver/topology/connection.go:299]: Error messages should generally use lowercase letters.

File: x/mongo/driver/topology/connection.go (hunk 3 above)
Comment (divjotarora) [inline x/mongo/driver/topology/connection.go:297]: Github doesn't let me comment in the right place because the code didn't change, but we do another `io.ReadFull` call below to read the rest of the message after we get the header. Do we need to change the message in that codepath too?

File: x/mongo/driver/topology/errors.go [hunk 5]
```diff
@@ -15,13 +19,43 @@ type ConnectionError struct {
 
 // Error implements the error interface.
 func (e ConnectionError) Error() string {
+	message := e.message
+	if e.init == true {
```
Comment (iwysiu) [inline x/mongo/driver/topology/errors.go:23]: this can just be "if e.init {"

File: x/mongo/driver/topology/errors.go [hunk 6]
```diff
@@ -15,13 +19,43 @@ type ConnectionError struct {
...
 // Unwrap returns the underlying error.
 func (e ConnectionError) Unwrap() error {
 	return e.Wrapped
 }
+
+// ServerSelectionError represents a Server Selection error.
+type ServerSelectionError struct {
+	Desc    description.Topology
+	Wrapped error
+	message string
```
Comment (divjotarora) [inline x/mongo/driver/topology/errors.go:44]: Is the `message` field ever used?

File: x/mongo/driver/topology/topology.go [hunk 7]
```diff
@@ -478,7 +474,7 @@ func (t *Topology) selectServerFromDescription(desc description.Topology,
 
 	suitable, err := selectionState.selector.SelectServer(desc, allowed)
 	if err != nil {
-		return nil, wrapServerSelectionError(err, t)
+		return nil, ServerSelectionError{Wrapped: err, Desc: t.Description()}
```
Comment (divjotarora) [inline x/mongo/driver/topology/topology.go:477]: The `SelectServer` call used `desc`, so the error should use `desc` rather than `t.Description()`. Remember that there's a background routine that updates server and topology descriptions, so `t.Description()` might return something different than `desc`, which would be confusing for users.

File: x/mongo/driver/topology/topology.go (hunk 1 above)
Comment (divjotarora) [inline x/mongo/driver/topology/topology.go:443]: I'm not 100% sure about this, but instinctively, I think we should be using the last seen topology description in the error message rather than fetching the current one from `t.Description()`. However, there's a chance that this is the first loop iteration so `current` hasn't been initialized.

What do you think about initializing `current` above the loop as `current := t.Description()` and then doing `return nil, ServerSelectionError{Wrapped: ctx.Error(), Desc: current}` here?

File: x/mongo/driver/topology/topology_test.go [hunk 8]
```diff
@@ -60,6 +60,10 @@ func TestServerSelection(t *testing.T) {
 	var selectError description.ServerSelectorFunc = func(description.Topology, []description.Server) ([]description.Server, error) {
 		return nil, errSelectionError
 	}
+	var errContextError = context.DeadlineExceeded
+	var selectContextError description.ServerSelectorFunc = func(description.Topology, []description.Server) ([]description.Server, error) {
+		return nil, errContextError
```
Comment (divjotarora) [inline x/mongo/driver/topology/topology_test.go:65]: nit: Do we need to declare `errContextError`? Can this be reduced to `return nil, context.DeadlineExceeded`?

File: x/mongo/driver/topology/topology_test.go [hunk 9]
```diff
@@ -160,7 +164,8 @@ func TestServerSelection(t *testing.T) {
 			t.Errorf("Timed out while trying to retrieve selected servers")
 		}
 
-		if err != context.Canceled {
+		want := ServerSelectionError{Wrapped: context.Canceled, Desc: topo.Description()}
+		if err.Error() != want.Error() {
```
Comment (divjotarora) [inline x/mongo/driver/topology/topology_test.go:168]: nit: This conditional can be changed to `assert.Equal(t, want, err, ...)`. Our `assert` library handles error comparisons by calling `Error()`: https://github.com/mongodb/mongo-go-driver/blob/a6cd2226c7836c1264a00b7101f751e95eeef570/internal/testutil/assert/assert.go#L19.

File: x/mongo/driver/topology/topology_test.go [hunk 10]
```diff
@@ -230,6 +235,30 @@ func TestServerSelection(t *testing.T) {
...
 	})
+	t.Run("Context Deadline Error", func(t *testing.T) {
+		desc := description.Topology{}
+		topo, err := New()
+		noerr(t, err)
+		subCh := make(chan description.Topology, 1)
+		subCh <- desc
+		resp := make(chan error)
+		timeout := make(chan time.Time)
+		go func() {
```
Comment (divjotarora) [inline x/mongo/driver/topology/topology_test.go:246]: I assume you copied this pattern of running a goroutine and selecting over a results channel and a `time.After` from the test above. This is fine, but I think you can use `assert.Soon` here to clean up the code a bit: https://github.com/mongodb/mongo-go-driver/blob/a6cd2226c7836c1264a00b7101f751e95eeef570/internal/testutil/assert/assert.go#L86.

File: x/mongo/driver/topology/topology_test.go [hunk 11]
```diff
@@ -230,6 +235,30 @@ func TestServerSelection(t *testing.T) {
...
+			state := newServerSelectionState(selectContextError, timeout)
+			ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
```
Comment (divjotarora) [inline x/mongo/driver/topology/topology_test.go:248]: nit: 1 second is a relatively long test execution time. Can this be `50*time.Millisecond` or does the time matter?

File: x/mongo/driver/topology/topology_test.go [hunk 12]
```diff
@@ -548,7 +577,6 @@ func TestTopology_String_Race(t *testing.T) {
 	}()
 
 	go func() {
-		topo.String()
```
Comment (divjotarora) [inline x/mongo/driver/topology/topology_test.go:551]: Why was this removed? This test asserts that the `topology.Topology.String()` function is not vulnerable to data races. To do that, this test launches two goroutines. One modifies `topo.servers` and the other concurrently calls `topo.String()`, which internally accesses `topo.servers`. Removing this call makes removes the core functionality of the test.

File: x/mongo/driver/topology/topology_test.go (hunk 10 above)
Comment (KritiRav) [reply x/mongo/driver/topology/topology_test.go]: done

File: x/mongo/driver/description/topology.go (hunk 2 above)
Comment (KritiRav) [reply x/mongo/driver/description/topology.go]: done

File: x/mongo/driver/topology/connection.go (hunk 3 above)
Comment (KritiRav) [reply x/mongo/driver/topology/connection.go]: done

Comment (KritiRav) [reply x/mongo/driver/topology/connection.go]: done

File: x/mongo/driver/topology/connection.go (hunk 4 above)
Comment (KritiRav) [reply x/mongo/driver/topology/connection.go]: done

File: x/mongo/driver/topology/errors.go (hunk 5 above)
Comment (KritiRav) [reply x/mongo/driver/topology/errors.go]: done

File: x/mongo/driver/topology/errors.go (hunk 6 above)
Comment (KritiRav) [reply x/mongo/driver/topology/errors.go]: the `message` field is not used, I've removed it

File: x/mongo/driver/topology/topology.go (hunk 7 above)
Comment (KritiRav) [reply x/mongo/driver/topology/topology.go]: done

File: x/mongo/driver/topology/topology_test.go (hunk 9 above)
Comment (KritiRav) [reply x/mongo/driver/topology/topology_test.go]: done

File: x/mongo/driver/topology/topology_test.go (hunk 11 above)
Comment (KritiRav) [reply x/mongo/driver/topology/topology_test.go]: done. Changed to 50*time.Millisecond

File: x/mongo/driver/topology/topology_test.go (hunk 8 above)
Comment (KritiRav) [reply x/mongo/driver/topology/topology_test.go]: done

File: x/mongo/driver/topology/topology.go (hunk 1 above)
Comment (KritiRav) [reply x/mongo/driver/topology/topology.go]: This makes sense, I've changed to use current as the last seen topology description

File: x/mongo/driver/topology/connection.go [hunk 13]
```diff
@@ -322,10 +324,14 @@ func (c *connection) readWireMessage(ctx context.Context, dst []byte) ([]byte, e
 	if err != nil {
 		// We closeConnection the connection because we don't know if there are other bytes left to read.
 		c.close()
+		message := "incomplete read of message header"
```
Comment (divjotarora) [inline x/mongo/driver/topology/connection.go:327]: This same code is repeated below, so let's refactor it into a function like `getReadErrrorMessage(err error) string`.

File: x/mongo/driver/topology/errors.go [hunk 14]
```diff
@@ -15,13 +19,39 @@ type ConnectionError struct {
 
 // Error implements the error interface.
 func (e ConnectionError) Error() string {
+	message := e.message
+	if e.init {
+		message += "error occured during connection handshake"
```
Comment (divjotarora) [inline x/mongo/driver/topology/errors.go:24]: If `e.message` is non-empty, this will concatenate a string to it without any spaces. It only works right now because we aren't setting a message for errors encountered during `connection.connect()`. How about doing:

```
fullMsg := "error occured during connection handshake
if message != "" {
    fullMsg = fmt.Sprintf("%s: %s", fullMsg, message)
}

message = fullMsg
```

This way, we get both pieces of information and don't have weird spacing if `e.message` is empty.

File: x/mongo/driver/topology/topology_test.go [hunk 15]
```diff
@@ -230,6 +232,29 @@ func TestServerSelection(t *testing.T) {
...
+		timeout := make(chan time.Time)
+		callback := func() {
+			state := newServerSelectionState(selectContextError, timeout)
+			ctx, cancel := context.WithTimeout(context.Background(), 50*time.Millisecond)
+			defer cancel()
+			_, err := topo.selectServerFromSubscription(ctx, subCh, state)
+			resp <- err
+			return
+		}
+		go func() {
```
Comment (divjotarora) [inline x/mongo/driver/topology/topology_test.go:251]: Why is this being called in a goroutine? `assert.Soon` will return as soon as the provided callback returns or the timeout elapses, so you should be able to call this synchronously.

File: x/mongo/driver/topology/topology_test.go [hunk 16]
```diff
@@ -230,6 +232,29 @@ func TestServerSelection(t *testing.T) {
 			t.Fatalf("did not receive error from server selection")
 		}
 	})
+	t.Run("Context Deadline Error", func(t *testing.T) {
+		desc := description.Topology{}
+		topo, err := New()
+		noerr(t, err)
+		subCh := make(chan description.Topology, 1)
+		subCh <- desc
+		resp := make(chan error)
```
Comment (divjotarora) [inline x/mongo/driver/topology/topology_test.go:241]: nit: You don't need a channel anymore. This was needed when the code below had to wait until an error or a timeout, which required selecting over two channels, but that part is now handled by `assert.Soon`. This can be:

```
var serverSelectionError error
callback := func() {
    // setup topology and server selection state
    // setup ctx
    _, serverSelectionError = topo.selectServerFromSubscription(ctx, subCh, state)
}

assert.Soon(callback, 100*time.Millisecond)
// assertions about serverSelectionError being correct
```

Note the use of `=` rather than `:=` in the callback to avoid shadowing.

File: x/mongo/driver/topology/topology_test.go [hunk 17]
```diff
@@ -230,6 +232,29 @@ func TestServerSelection(t *testing.T) {
...
+		timeout := make(chan time.Time)
+		callback := func() {
+			state := newServerSelectionState(selectContextError, timeout)
```
Comment (divjotarora) [inline x/mongo/driver/topology/topology_test.go:244]: Using the `selectContextError` selector is fine here, but it's more realistic for this error to happen in the actual server selection loop rather than the selector. You can use the the `Cancel` test above for inspiration. It uses the `selectNone` selector, so server selection will loop and then timeout after 50ms because that's your context timeout. If you do this, you can also remove the `selectContextError` selector if it's unused.

File: x/mongo/driver/topology/topology_test.go [hunk 18]
```diff
@@ -160,9 +163,8 @@ func TestServerSelection(t *testing.T) {
 			t.Errorf("Timed out while trying to retrieve selected servers")
 		}
 
-		if err != context.Canceled {
-			t.Errorf("Incorrect error received. got %v; want %v", err, context.Canceled)
-		}
+		want := ServerSelectionError{Wrapped: context.Canceled, Desc: desc}
+		assert.Equal(t, err, want, "Incorrect error received. got %v; want %v", err, want)
```
Comment (divjotarora) [inline x/mongo/driver/topology/topology_test.go:167]: I'd like this test to assert that the `Unwrap` function works as expected. You can do that using the `errors.Is` function (https://blog.golang.org/go1.13-errors). An example of this in our existing tests: https://github.com/mongodb/mongo-go-driver/blob/master/mongo/integration/errors_test.go#L36. Note that you'll need to add a build flag to the top of this file so it only runs tests if compiled with Go 1.13 or higher: https://github.com/mongodb/mongo-go-driver/blob/master/mongo/integration/errors_test.go#L7.

File: x/mongo/driver/topology/topology_test.go (hunk 12 above)
Comment (divjotarora) [reply x/mongo/driver/topology/topology_test.go]: If your compiler is complaining about it, can you try changing it to `_ = topo.String()`?

File: x/mongo/driver/topology/connection.go (hunk 13 above)
Comment (KritiRav) [reply x/mongo/driver/topology/connection.go:327]: I'm not sure how necessary it would be to refactor this into a function like `getReadErrrorMessage(err error) string`. The original message is slightly different (`incomplete read of message header` vs `incomplete read of full message`), so `getReadErrorMessage` would basically just be one if statement

Comment (divjotarora) [reply x/mongo/driver/topology/connection.go:327]: Ah, never mind then. This is fine. Resolving.
